File: solutions/haiku-1/haiku/go/wordfreq.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strings"
)

func main() {
	scanner := bufio.NewScanner(os.Stdin)

	// Read all input and count words
	wordCount := make(map[string]int)

	for scanner.Scan() {
		line := scanner.Text()
		words := extractWords(line)
		for _, word := range words {
			wordCount[word]++
		}
	}

	// Create a slice of word-count pairs
	type wordCountPair struct {
		word  string
		count int
	}

	pairs := make([]wordCountPair, 0, len(wordCount))
	for word, count := range wordCount {
		pairs = append(pairs, wordCountPair{word, count})
	}

	// Sort by count descending, then by word ascending
	sort.Slice(pairs, func(i, j int) bool {
		if pairs[i].count != pairs[j].count {
			return pairs[i].count > pairs[j].count
		}
		return pairs[i].word < pairs[j].word
	})

	// Print results
	for _, p := range pairs {
		fmt.Printf("%s %d\n", p.word, p.count)
	}
}

func extractWords(text string) []string {
	var words []string
	var currentWord strings.Builder

	for _, r := range text {
		// Check if it's an ASCII letter (A-Z or a-z)
		if (r >= 'A' && r <= 'Z') || (r >= 'a' && r <= 'z') {
			// Convert to lowercase
			if r >= 'A' && r <= 'Z' {
				currentWord.WriteRune(r + 32)
			} else {
				currentWord.WriteRune(r)
			}
		} else {
			// Non-letter character ends the current word
			if currentWord.Len() > 0 {
				words = append(words, currentWord.String())
				currentWord.Reset()
			}
		}
	}

	// Don't forget the last word
	if currentWord.Len() > 0 {
		words = append(words, currentWord.String())
	}

	return words
}

File: solutions/haiku-1/haiku/java/wordfreq.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        Map<String, Integer> wordCount = new HashMap<>();

        // Read from stdin
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        String line;
        while ((line = reader.readLine()) != null) {
            // Extract words (maximal runs of ASCII letters)
            int i = 0;
            while (i < line.length()) {
                char c = line.charAt(i);
                // Check if character is an ASCII letter (a-z or A-Z)
                if ((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z')) {
                    StringBuilder word = new StringBuilder();
                    // Collect all consecutive ASCII letters
                    while (i < line.length()) {
                        char ch = line.charAt(i);
                        if ((ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z')) {
                            word.append(ch);
                            i++;
                        } else {
                            break;
                        }
                    }
                    // Convert to lowercase and count
                    String wordStr = word.toString().toLowerCase();
                    wordCount.put(wordStr, wordCount.getOrDefault(wordStr, 0) + 1);
                } else {
                    i++;
                }
            }
        }

        // Sort by count descending, then by word ascending
        List<Map.Entry<String, Integer>> entries = new ArrayList<>(wordCount.entrySet());
        entries.sort((a, b) -> {
            // First compare by count (descending)
            if (!a.getValue().equals(b.getValue())) {
                return b.getValue().compareTo(a.getValue());
            }
            // Then compare by word (ascending)
            return a.getKey().compareTo(b.getKey());
        });

        // Output
        for (Map.Entry<String, Integer> entry : entries) {
            System.out.println(entry.getKey() + " " + entry.getValue());
        }
    }
}
